### PR TITLE
Sort communities by currently active tab

### LINKED ISSUE #17337: The communities sorted randomly on the list of communities page.
### Steps:
1. User_A: Creates at least 3-4 communities -> share them to User_B
2. User_B: Joins them
3. User_B: Check the list of communities in the 'joined' tab

### Actual result:
The communities are sorted randomly
image

### Expected result:
- The communities are sorted in order of joined
https://discord.com/channels/624634427930312714/842696205213368331/1151526646675947651

- In the 'pending' tab, the communities should be sorted in the order they were requested.
- For 'opened' communities, they should be sorted based on the order they were shared.

### PR DESCRIPTION
**Pull request by ibrkhalil**
fixes #17337

### Summary

Sorts communities by last-opened-at property, And falls back to joined-at property if last-opened-at doesn't exist.

status: ready

### CODE REVIEW

File: src/status_im/contexts/communities/home/view.cljs [hunk 1]
```diff
@@ -100,7 +100,10 @@
              :header                            [common.header-spacing/view]
              :render-fn                         item-render
              :style                             {:margin-top -1}
-             :data                              selected-items
+             :data                              (sort-by (fn [{:keys [last-opened-at joined-at]}]
```
Comment (cammellos) [inline src/status_im/contexts/communities/home/view.cljs]: should be done in subscription

File: src/status_im/contexts/communities/events.cljs [hunk 2]
```diff
@@ -21,15 +21,18 @@
...
                                 (or (= type constants/community-token-permission-can-view-channel)
                                     (=
                                      type
-                                     constants/community-token-permission-can-view-and-post-channel))))]
-      {:db (assoc-in db [:communities id] community)
+                                     constants/community-token-permission-can-view-and-post-channel))))
+          previous-last-opened-at (get-in db [:communities id :last-opened-at])]
+      {:db (assoc-in db
+            [:communities id]
+            (assoc community :last-opened-at (max last-opened-at previous-last-opened-at)))
```
Comment (ibrkhalil) [inline src/status_im/contexts/communities/events.cljs:36]: This is a fix to handle a race condition that happens between fetching the last community details and updating the last opened at key when opening a community.

File: src/status_im/subs/communities.cljs [hunk 3]
```diff
@@ -107,7 +107,10 @@
                                              pending? (update acc :pending conj community)
                                              :else    (update acc :opened conj community))))
                                        {:joined [] :pending [] :opened []}
-                                       (vals communities))]
+                                       (sort-by (fn [{:keys [last-opened-at joined-at]}]
```
Comment (shivekkhurana) [inline src/status_im/subs/communities.cljs]: This reducer seems to be going out of control. Can you simplify it  ?
```clj
(reduce (fn [acc community]
          (let [joined?      (:joined community)
                community-id (:id community)
                pending?     (boolean (get requests community-id))]
            (cond
              joined?  (update acc :joined conj community)
              pending? (update acc :pending conj community)
              :else    (update acc :opened conj community))))
        {:joined [] :pending [] :opened []}
        (sort-by (fn [{:keys [last-opened-at joined-at]}]
                   (or last-opened-at joined-at))
                 #(compare %2 %1)
                 (vals communities)))
```

I think we can get rid of the reducer completely. Something like this should be doable:

```clj
(->> communities
         vals
         (sort-by ...)
         (group-by ...)
         (merge-with ...))
```
Links to relavent functions:
- https://clojuredocs.org/clojure.core/group-by
- https://clojuredocs.org/clojure.core/merge-with

**Review by shivekkhurana — CHANGES_REQUESTED** [review]

File: src/status_im/subs/communities.cljs (hunk 3 above)
Comment (ibrkhalil) [reply src/status_im/subs/communities.cljs]: Done, And thank you for your help on this Shivek!

File: src/status_im/subs/communities.cljs [hunk 4]
```diff
@@ -88,6 +88,15 @@
...
 
+(defn- group-communities-by-status
+  [requests {:keys [id] :as community}]
+  (let [joined?  (:joined community)
+        pending? (boolean (get requests id))]
+    (cond
+      joined?  :joined
+      pending? :pending
+      :else    :opened)))
+
```
Comment (shivekkhurana) [inline src/status_im/subs/communities.cljs:109]: A more crisp version:
```clj
(defn- group-communities-by-status
  [requests {:keys [id] :as community}]
  (cond
    (:joined community)         :joined
    (boolean (get requests id)) :pending
    :else                       :opened))
```

**Review by shivekkhurana — APPROVED** [review]
image

File: src/status_im/subs/communities.cljs (hunk 4 above)
Comment (ibrkhalil) [reply src/status_im/subs/communities.cljs:109]: Done

**Comment by mariia-skrypnyk** [comment]
**ISSUE 1: Opened tab isn't sorted by the last opened if user opens communities from links in 1 to 1 chat**

**Preconditions:** User B (desktop) created 3 communities with required approve


**Steps:** 

1. Login as user A
2. Add User B to contacts (desktop user)
3. As User B share all 3 communities
4. As User A open links to every community

**Actual result:** Most opened on the bottom

https://github.com/status-im/status-mobile/assets/147824399/59ce3891-e064-4850-96e6-463820439834


**Expected result:** Most opened on top

Devices: iOS/Android

**Comment by mariia-skrypnyk** [comment]
**ISSUE 2: Most recent joined are not on the top**

**Preconditions:** User B (desktop) created 3 communities with required approve

**Steps:**

1. Login as user A
2. Add User B to contacts (desktop user)
3. As User B share all 3 communities
4. As User A request to join all 3 communities
5. As User B accept request by request from the first request to the last

**Actual result:**

1) User B accepts requests in such sequence:
Pink
Greener
Red
2) User B sees such list in his Joined tab:
Pink
Greener
Red

So there is no "Most recent joined are not on the top" logic here

**Expected result:** Most recent joined on the top

Red
Greener
Pink

**Comment by ibrkhalil** [comment]
@mariia-skrypnyk Both issues should be fixed now, Can you please take another look?

**Comment by mariia-skrypnyk** [comment]
@ibrkhalil please, do rebase.

**Comment by ibrkhalil** [comment]
> @ibrkhalil please, do rebase.

Done

**Comment by mariia-skrypnyk** [comment]
Hi @ibrkhalil !

Issue 2 is fixed!
Issue 1 still have some points:

According to https://github.com/status-im/status-mobile/issues/17337#issuecomment-1851952622 we need such implementation:

1) Opened tab SHOULD show us communities we opened from last opened on top.
Also Opened tab SHOULD include communities that we Joined and NOT Joined too. 
Now we have a case when opened Joined communities are not shown in Opened tab. Please look at the video:

https://github.com/status-im/status-mobile/assets/147824399/649a63db-bbce-4e7a-bd01-98da3db86593
Actual result: I've open joined communities but they are not shown in Opened tab.

2) The second issue also is described in video above: when you open communities from Joined tab they become sorted 
NOT by order of joining but by order of opening in Joined tab. BUT they should be ordered by order of joining:
Screenshot 2024-03-04 at 14 39 26

**Comment by ibrkhalil** [comment]
> Hi @ibrkhalil !
> 
> Issue 2 is fixed! Issue 1 still have some points:
> 
> According to [#17337 (comment)](https://github.com/status-im/status-mobile/issues/17337#issuecomment-1851952622) we need such implementation:
> 
> 1. Opened tab SHOULD show us communities we opened from last opened on top.
>    Also Opened tab SHOULD include communities that we Joined and NOT Joined too.
>    Now we have a case when opened Joined communities are not shown in Opened tab. Please look at the video:
> 
> https://github.com/status-im/status-mobile/assets/147824399/649a63db-bbce-4e7a-bd01-98da3db86593 Actual result: I've open joined communities but they are not shown in Opened tab.
> 
> 2. The second issue also is described in video above: when you open communities from Joined tab they become sorted
>    NOT by order of joining but by order of opening in Joined tab. BUT they should be ordered by order of joining:
> 
> Screenshot 2024-03-04 at 14 39 26

Should be fixed now.

**Comment by mariia-skrypnyk** [comment]
Hi @ibrkhalil !

Thanks for such a long sorting journey!
Now all sorting requirements are covered and PR can be merged!
Great job!
Failed e2e are not related to your fixes.

**Pull request merged**